samza-sql-api/src/main/java/org/apache/samza/sql/interfaces/udfs/ScalarUdf.java

```diff
@@ -0,0 +1,33 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.interfaces.udfs;
+
+public abstract class ScalarUdf {
+
+  private final ScalarUdfContext udfContext;
+
+  public ScalarUdf() {
+    udfContext = ScalarUdfContext.getInstance();
```

Comment:
Generally this is an antipattern as it complicates testing and tightly couples code. It looks like you're loading UDFs later by reflection. Would it be possible to (potentially optionally) pass context in via reflection to the execute method instead of adding a hard dependency here?

Reply:
Thanks for flagging this. Yup i am not particular pleased with this solution too, The problem i am facing here is.

Since we use calcite as execution engine, It is the calcite which actually instantiates the UDF classes and calls the execute method directly. We just pass the Method that needs to be instantiated to calcite [here](https://github.com/srinipunuru/samza/blob/93f1e78ddc2878aad537d1375f2879c0d9d18394/samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java#L111
).  Do you see any technique we could use to set the context in this scenario? 

I will add this as one of the requests to the calcite community to see if there is a way to hookup context to the UDF objects that calcite creates.

Reply:
That definitely constrains your options. Checking with the Calcite community makes sense. Essentially we'd like to parameterize the UDF with values from the runtime environment.